File: blaise-svg/src/main/java/com/googlecode/blaisemath/graphics/svg/SvgElementGraphicConverter.java
package com.googlecode.blaisemath.graphics.svg;

/*
 * #%L
 * BlaiseSVG
 * --
 * Copyright (C) 2014 - 2019 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.google.common.base.Converter;
import com.google.common.base.Strings;
import com.googlecode.blaisemath.graphics.AnchoredIcon;
import com.googlecode.blaisemath.graphics.AnchoredImage;
import com.googlecode.blaisemath.graphics.AnchoredText;
import com.googlecode.blaisemath.graphics.core.Graphic;
import com.googlecode.blaisemath.graphics.core.GraphicComposite;
import com.googlecode.blaisemath.graphics.core.PrimitiveArrayGraphicSupport;
import com.googlecode.blaisemath.graphics.core.PrimitiveGraphicSupport;
import com.googlecode.blaisemath.graphics.swing.JGraphicComponent;
import com.googlecode.blaisemath.graphics.swing.JGraphics;
import com.googlecode.blaisemath.graphics.swing.LabeledShapeGraphic;
import com.googlecode.blaisemath.graphics.swing.PanAndZoomHandler;
import com.googlecode.blaisemath.graphics.swing.TextRenderer;
import com.googlecode.blaisemath.graphics.swing.WrappedTextRenderer;
import com.googlecode.blaisemath.style.*;
import com.googlecode.blaisemath.svg.*;
import com.googlecode.blaisemath.svg.SvgElement;
import com.googlecode.blaisemath.util.Colors;
import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.Shape;
import java.awt.geom.Line2D;
import java.awt.geom.Point2D;
import java.awt.geom.RectangularShape;
import java.util.Map;
import java.util.Map.Entry;
import java.util.logging.Level;
import java.util.logging.Logger;

/** 
 * Adapter for converting SVG objects to/from Blaise {@link PrimitiveGraphicSupport} objects.
 * The conversion is imperfect in this implementation. Supported SVG types include:
 * <ul>
 *   <li>{@link SvgRectangle}, {@link SvgEllipse}, {@link SvgCircle}, {@link SvgPolygon}</li>
 *   <li>{@link SvgLine}, {@link SvgPolyline}</li>
 *   <li>{@link SvgPath}</li>
 *   <li>{@link SvgImage}</li>
 *   <li>{@link SvgText}</li>
 *   <li>{@link SvgGroup}</li>
 * </ul>
 * This adapter also converts {@link SvgGroup} to/from {@link GraphicComposite}.
 * 
 * @author Elisha Peterson
 */
public class SvgElementGraphicConverter extends Converter<SvgElement, Graphic<Graphics2D>> {
    
    private static final Logger LOG = Logger.getLogger(SvgElementGraphicConverter.class.getName());
    private static final SvgElementGraphicConverter INST = new SvgElementGraphicConverter();

    /**
     * Get global instance of the converter.
     * @return instance
     */
    public static Converter<SvgElement, Graphic<Graphics2D>> getInstance() {
        return INST;
    }

    /**
     * Convert a graphic component to an SVG object, including a view box.
     * @param compt component to convert
     * @return result
     */
    public static SvgRoot componentToSvg(JGraphicComponent compt) {
        SvgRoot root = new SvgRoot();
        root.setWidth(compt.getWidth());
        root.setHeight(compt.getHeight());
        root.setViewBoxAsRectangle(PanAndZoomHandler.getLocalBounds(compt));
        root.getStyle().put("background", Colors.encode(compt.getBackground()));
        root.getStyle().put(Styles.FONT_SIZE, Styles.DEFAULT_TEXT_STYLE.get(Styles.FONT_SIZE));
        SvgGroup group = (SvgGroup) SvgElementGraphicConverter.getInstance().reverse()
                .convert(compt.getGraphicRoot());
        group.getElements().forEach(root::addElement);
        return root;
    }
    
    /**
     * Convert an {@link SvgElement} to a {@link Graphic}. The resulting object will
     * be a {@link GraphicComposite} if the argument is an instance of {@link SvgGroup}
     * or {@link SvgRoot}, and otherwise a {@link PrimitiveGraphicSupport}. In the case of
     * a group, recursive calls are made to convert all elements in the group.
     * 
     * @param sh the element to convert
     * @return the corresponding graphic
     */
    @Override
    public Graphic<Graphics2D> doForward(SvgElement sh) {
        Graphic<Graphics2D> prim = null;
        AttributeSet style = aggregateStyle(sh);
        if (sh instanceof SvgRectangle) {
            RectangularShape rsh = SvgRectangle.shapeConverter().convert((SvgRectangle) sh);
            prim = JGraphics.shape(rsh, style);
        } else if (sh instanceof SvgEllipse) {
            Shape rsh = SvgEllipse.shapeConverter().convert((SvgEllipse) sh);
            prim = JGraphics.shape(rsh, style);
        } else if (sh instanceof SvgCircle) {
            Shape rsh = SvgCircle.shapeConverter().convert((SvgCircle) sh);
            prim = JGraphics.shape(rsh, style);
        } else if (sh instanceof SvgPolygon) {
            Shape rsh = SvgPolygon.shapeConverter().convert((SvgPolygon) sh);
            prim = JGraphics.shape(rsh, style);
        } else if (sh instanceof SvgLine) {
            Line2D line = SvgLine.shapeConverter().convert((SvgLine) sh);
            prim = JGraphics.path(line, style);
        } else if (sh instanceof SvgPolyline) {
            Shape rsh = SvgPolyline.shapeConverter().convert((SvgPolyline) sh);
            prim = JGraphics.shape(rsh, style);
        } else if (sh instanceof SvgPath) {
            Shape rsh = SvgPath.shapeConverter().convert((SvgPath) sh);
            prim = JGraphics.shape(rsh, style);
        } else if (sh instanceof SvgImage) {
            AnchoredImage img = SvgImage.imageConverter().convert((SvgImage) sh);
            prim = JGraphics.image(img);
            prim.setMouseDisabled(true);
        } else if (sh instanceof SvgText) {
            AnchoredText text = SvgText.textConverter().convert((SvgText) sh);
            prim = JGraphics.text(text, style);
            prim.setMouseDisabled(true);
        } else if (sh instanceof SvgGroup || sh instanceof SvgRoot) {
            prim = new GraphicComposite<>();
            ((GraphicComposite)prim).setStyle(style);
            for (SvgElement el : ((SvgGroup)sh).getElements()) {
                ((GraphicComposite<Graphics2D>)prim).addGraphic(doForward(el));
            }
        } else {
            throw new IllegalStateException("Unexpected SVG element: "+sh);
        }
        prim.setDefaultTooltip(sh.getId());
        return prim;
    }

    private AttributeSet aggregateStyle(SvgElement element) {
        AttributeSet shapeStyle = element.getStyle();
        AttributeSet res = shapeStyle == null 
                ? new AttributeSet()
                : shapeStyle.copy();
        
        Map<String, Object> attr = element.getOtherAttributes();
        if (attr != null) {
            for (Entry<String, Object> en : attr.entrySet()) {
                Object val = new AttributeSetCoder().decode((String) en.getValue());
                res.put(en.getKey(), val);
            }
        }
        if (element.getId() != null) {
            res.put(Styles.ID, element.getId());
        }
        AttributeSetAdapter.updateColorFields(res);
        return res;
    }

    @Override
    public SvgElement doBackward(Graphic<Graphics2D> v) {
        return graphicToSvg(v);
    }
    
    //region UTILITIES
    
    /** Converts a graphic element to an SVG element */
    private static SvgElement graphicToSvg(Graphic<Graphics2D> v) {
        SvgElement res = null;
        if (v instanceof LabeledShapeGraphic) {
            res = labeledShapeToSvg((LabeledShapeGraphic<Graphics2D>) v);
        } else if (v instanceof PrimitiveGraphicSupport) {
            PrimitiveGraphicSupport pgs = (PrimitiveGraphicSupport) v;
            res = primitiveStyleToSvg(pgs.getPrimitive(), v.getStyle(), pgs.getRenderer());
        } else if (v instanceof GraphicComposite) {
            res = compositeToSvg((GraphicComposite<Graphics2D>) v);
        } else if (v instanceof PrimitiveArrayGraphicSupport) {
            res = primitiveArrayToSvg((PrimitiveArrayGraphicSupport) v);
        } else {
            throw new IllegalArgumentException("Graphic conversion not supported for "+v.getClass());
        }
        String id = v.renderStyle().getString(Styles.ID, null);
        if (id != null && res != null) {
            res.setId(id);
        }
        if (res != null && res.getStyle() != null && res.getStyle().getAttributes().isEmpty()) {
            res.setStyle(null);
        }
        if (res != null && res.getStyle() != null) {
            for (String c : res.getStyle().getAttributes()) {
                Object col = res.getStyle().get(c);
                if (col instanceof Color) {
                    res.getStyle().put(c, Colors.alpha((Color) col, 255));
                }
            }
        }
        return res;
    }

    /** Converts a blaise composite to an SVG group */
    private static SvgElement compositeToSvg(GraphicComposite<Graphics2D> gc) {
        SvgGroup grp = new SvgGroup();
        if (gc.getStyle() != null) {
            grp.setStyle(AttributeSet.create(gc.getStyle().getAttributeMap()));
        }
        for (Graphic<Graphics2D> g : gc.getGraphics()) {
            try {
                SvgElement el = graphicToSvg(g);
                if (el != null) {
                    grp.addElement(el);
                } else {
                    LOG.log(Level.WARNING, "Null graphic for {0}", g);
                }
            } catch (IllegalArgumentException x) {
                LOG.log(Level.WARNING, "Graphic not added to result", x);
            }
        }
        return grp;
    }

    /** Converts a blaise array graphic to SVG group */
    private static SvgElement primitiveArrayToSvg(PrimitiveArrayGraphicSupport pags) {
        SvgGroup grp = new SvgGroup();
        grp.setStyle(pags.renderStyle().flatCopy());
        for (Object o : pags.getPrimitive()) {
            grp.addElement(primitiveStyleToSvg(o, pags.renderStyle().flatCopy(), pags.getRenderer()));
        }
        return grp;
    }

    /** Creates an SVG element from given primitive/style */
    private static SvgElement primitiveStyleToSvg(Object primitive, AttributeSet sty, Renderer rend) {
        if (primitive instanceof Shape) {
            return SvgElements.create(null, (Shape) primitive, sty);
        } else if (primitive instanceof AnchoredText) {
            return SvgElements.create(null, (AnchoredText) primitive, sty, rend);
        } else if (primitive instanceof AnchoredImage) {
            return SvgElements.create(null, (AnchoredImage) primitive, sty);
        } else if (primitive instanceof AnchoredIcon) {
            return SvgElements.create(null, (AnchoredIcon) primitive, sty);
        } else if (primitive instanceof Point2D) {
            return SvgElements.create(null, (Point2D) primitive, sty);
        } else {
            throw new IllegalArgumentException("Graphic conversion not supported for primitive "+primitive);
        }
    }

    /** Converts a labeled shape to svg */
    private static SvgElement labeledShapeToSvg(LabeledShapeGraphic<Graphics2D> gfc) {
        SvgElement shape = primitiveStyleToSvg(gfc.getPrimitive(), gfc.renderStyle().flatCopy(), gfc.getRenderer());
        SvgElement text = labelToSvg(gfc);
        return text == null ? shape : SvgGroup.create(shape, text);
    }

    /** Generates element for object label */
    private static SvgElement labelToSvg(LabeledShapeGraphic<Graphics2D> gfc) {
        ObjectStyler<Graphics2D> styler = gfc.getObjectStyler();
        if (styler == null) {
            return null;
        }
        String label = styler.label(gfc.getSourceObject());
        AttributeSet style = styler.labelStyle(gfc.getSourceObject());
        if (Strings.isNullOrEmpty(label) || style == null) {
            return null;
        }
        Renderer<AnchoredText, Graphics2D> textRend = gfc.getTextRenderer();
        if (textRend instanceof WrappedTextRenderer) {
            return SvgElements.createWrappedText(label, style, LabeledShapeGraphic.wrappedLabelBounds(gfc.getPrimitive()));
        } else if (textRend instanceof TextRenderer) {
            return primitiveStyleToSvg(new AnchoredText(label), style.flatCopy(), textRend);
        } else {
            LOG.log(Level.WARNING, "Unsupported text renderer: {0}", textRend);
            return null;
        }
    }
    
    //endregion
    
}

File: blaise-svg-kt/src/main/java/com/googlecode/blaisemath/graphics/svg/SvgElementGraphicConverter.kt
package com.googlecode.blaisemath.graphics.svg
/*
 * #%L
 * BlaiseSvg
 * --
 * Copyright (C) 2014 - 2021 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import com.google.common.base.Converter
import com.google.common.base.Strings
import com.google.common.graph.ElementOrder
import com.google.common.graph.EndpointPair
import com.google.common.graph.GraphBuilder
import com.google.common.graph.Graphs
import com.google.common.graph.ImmutableGraph
import com.google.common.graph.MutableGraph
import com.googlecode.blaisemath.app.ApplicationMenuConfig
import com.googlecode.blaisemath.graph.ContractedGraphTest
import com.googlecode.blaisemath.graph.GraphComponents
import com.googlecode.blaisemath.graph.GraphGenerator
import com.googlecode.blaisemath.graph.GraphMetric
import com.googlecode.blaisemath.graph.GraphMetrics
import com.googlecode.blaisemath.graph.GraphNodeMetric
import com.googlecode.blaisemath.graph.GraphNodeStats
import com.googlecode.blaisemath.graph.GraphServices
import com.googlecode.blaisemath.graph.GraphSubsetMetric
import com.googlecode.blaisemath.graph.GraphUtilsTest
import com.googlecode.blaisemath.graph.IterativeGraphLayout
import com.googlecode.blaisemath.graph.IterativeGraphLayoutState
import com.googlecode.blaisemath.graph.NodeInGraph
import com.googlecode.blaisemath.graph.OptimizedGraph
import com.googlecode.blaisemath.graph.StaticGraphLayout
import com.googlecode.blaisemath.graph.SubgraphTest
import com.googlecode.blaisemath.graph.app.AnimationUtils
import com.googlecode.blaisemath.graph.app.GraphApp
import com.googlecode.blaisemath.graph.app.GraphAppCanvas
import com.googlecode.blaisemath.graph.app.GraphAppFrameView
import com.googlecode.blaisemath.graph.app.MetricScaler
import com.googlecode.blaisemath.graph.generate.AbstractGraphGenerator
import com.googlecode.blaisemath.graph.generate.CompleteGraphGenerator
import com.googlecode.blaisemath.graph.generate.CycleGraphGenerator
import com.googlecode.blaisemath.graph.generate.DefaultGeneratorParameters
import com.googlecode.blaisemath.graph.generate.DegreeDistributionGenerator
import com.googlecode.blaisemath.graph.generate.DegreeDistributionGenerator.DegreeDistributionParameters
import com.googlecode.blaisemath.graph.generate.EdgeCountGenerator
import com.googlecode.blaisemath.graph.generate.EdgeLikelihoodGenerator
import com.googlecode.blaisemath.graph.generate.EdgeLikelihoodGenerator.EdgeLikelihoodParameters
import com.googlecode.blaisemath.graph.generate.EmptyGraphGenerator
import com.googlecode.blaisemath.graph.generate.ExtendedGeneratorParameters
import com.googlecode.blaisemath.graph.generate.GraphGenerators
import com.googlecode.blaisemath.graph.generate.GraphGrowthRule
import com.googlecode.blaisemath.graph.generate.GraphSeedRule
import com.googlecode.blaisemath.graph.generate.HopGrowthRule
import com.googlecode.blaisemath.graph.generate.PreferentialAttachmentGenerator
import com.googlecode.blaisemath.graph.generate.PreferentialAttachmentGenerator.PreferentialAttachmentParameters
import com.googlecode.blaisemath.graph.generate.ProximityGraphGenerator.ProximityGraphParameters
import com.googlecode.blaisemath.graph.generate.StarGraphGenerator
import com.googlecode.blaisemath.graph.generate.WattsStrogatzGenerator
import com.googlecode.blaisemath.graph.generate.WattsStrogatzGenerator.WattsStrogatzParameters
import com.googlecode.blaisemath.graph.generate.WheelGraphGenerator
import com.googlecode.blaisemath.graph.layout.CircleLayout
import com.googlecode.blaisemath.graph.layout.CircleLayout.CircleLayoutParameters
import com.googlecode.blaisemath.graph.layout.GraphLayoutConstraints
import com.googlecode.blaisemath.graph.layout.GraphLayoutManager
import com.googlecode.blaisemath.graph.layout.IterativeGraphLayoutManager
import com.googlecode.blaisemath.graph.layout.IterativeGraphLayoutService
import com.googlecode.blaisemath.graph.layout.LayoutRegion
import com.googlecode.blaisemath.graph.layout.PositionalAddingLayout
import com.googlecode.blaisemath.graph.layout.RandomBoxLayout
import com.googlecode.blaisemath.graph.layout.RandomBoxLayout.BoxLayoutParameters
import com.googlecode.blaisemath.graph.layout.SpringLayoutParameters
import com.googlecode.blaisemath.graph.layout.SpringLayoutPerformanceTest
import com.googlecode.blaisemath.graph.layout.SpringLayoutState
import com.googlecode.blaisemath.graph.layout.StaticSpringLayout
import com.googlecode.blaisemath.graph.layout.StaticSpringLayout.StaticSpringLayoutParameters
import com.googlecode.blaisemath.graph.metrics.AbstractGraphMetric
import com.googlecode.blaisemath.graph.metrics.AbstractGraphNodeMetric
import com.googlecode.blaisemath.graph.metrics.AdditiveSubsetMetricTest
import com.googlecode.blaisemath.graph.metrics.BetweenCentrality
import com.googlecode.blaisemath.graph.metrics.BetweenCentralityTest
import com.googlecode.blaisemath.graph.metrics.ClosenessCentrality
import com.googlecode.blaisemath.graph.metrics.ClosenessCentralityTest
import com.googlecode.blaisemath.graph.metrics.ClusteringCoefficient
import com.googlecode.blaisemath.graph.metrics.CooperationMetric
import com.googlecode.blaisemath.graph.metrics.DecayCentrality
import com.googlecode.blaisemath.graph.metrics.EigenCentrality
import com.googlecode.blaisemath.graph.metrics.EigenCentralityTest
import com.googlecode.blaisemath.graph.metrics.GraphCentrality
import com.googlecode.blaisemath.graph.metrics.GraphCentralityTest
import com.googlecode.blaisemath.graph.metrics.GraphDiameter
import com.googlecode.blaisemath.graph.metrics.SubsetMetrics
import com.googlecode.blaisemath.graph.metrics.SubsetMetrics.AdditiveSubsetMetric
import com.googlecode.blaisemath.graph.metrics.SubsetMetrics.ContractiveSubsetMetric
import com.googlecode.blaisemath.graph.metrics.SubsetMetricsTest
import com.googlecode.blaisemath.graph.test.DynamicGraphTestFrame
import com.googlecode.blaisemath.graph.test.GraphTestFrame
import com.googlecode.blaisemath.graph.test.MyTestGraph
import com.googlecode.blaisemath.graph.util.Matrices
import com.googlecode.blaisemath.graph.view.GraphComponent
import com.googlecode.blaisemath.graph.view.VisualGraph
import com.googlecode.blaisemath.graph.view.WeightedEdgeStyler
import com.googlecode.blaisemath.graphics.AnchoredIcon
import com.googlecode.blaisemath.graphics.AnchoredImage
import com.googlecode.blaisemath.graphics.AnchoredText
import com.googlecode.blaisemath.graphics.core.Graphic
import com.googlecode.blaisemath.graphics.core.GraphicComposite
import com.googlecode.blaisemath.graphics.core.PrimitiveArrayGraphicSupport
import com.googlecode.blaisemath.graphics.core.PrimitiveGraphicSupport
import com.googlecode.blaisemath.graphics.editor.AttributeSetPropertyModelTestFrame
import com.googlecode.blaisemath.graphics.swing.*
import com.googlecode.blaisemath.graphics.swing.render.TextRenderer
import com.googlecode.blaisemath.graphics.swing.render.WrappedTextRenderer
import com.googlecode.blaisemath.graphics.testui.ContextMenuTestFrame
import com.googlecode.blaisemath.graphics.testui.HelloWorldTest
import com.googlecode.blaisemath.graphics.testui.PanAndZoomTestFrame
import com.googlecode.blaisemath.graphics.testui.SelectionTestFrame
import com.googlecode.blaisemath.graphics.testui.TooltipTestFrame
import com.googlecode.blaisemath.style.AttributeSet
import com.googlecode.blaisemath.style.AttributeSetCoder
import com.googlecode.blaisemath.style.Renderer
import com.googlecode.blaisemath.style.Styles
import com.googlecode.blaisemath.style.xml.AttributeSetAdapter
import com.googlecode.blaisemath.svg.*
import com.googlecode.blaisemath.svg.HelloWorldSvg
import com.googlecode.blaisemath.svg.SvgPathTest
import com.googlecode.blaisemath.svg.SvgRootTest
import com.googlecode.blaisemath.svg.SvgTool
import com.googlecode.blaisemath.test.AssertUtils
import com.googlecode.blaisemath.ui.PropertyActionPanel
import junit.framework.TestCase
import org.apache.commons.math.stat.descriptive.SummaryStatistics
import org.junit.BeforeClass
import java.awt.Color
import java.awt.Graphics2D
import java.awt.Shape
import java.awt.geom.Line2D
import java.awt.geom.Point2D
import java.awt.geom.RectangularShape
import java.util.function.Consumer
import java.util.logging.Level
import java.util.logging.Logger

/**
 * Adapter for converting Svg objects to/from Blaise [PrimitiveGraphicSupport] objects.
 * The conversion is imperfect in this implementation. Supported Svg types include:
 *
 *  * [SvgRectangle], [SvgEllipse], [SvgCircle], [SvgPolygon]
 *  * [SvgLine], [SvgPolyline]
 *  * [SvgPath]
 *  * [SvgImage]
 *  * [SvgText]
 *  * [SvgGroup]
 *
 * This adapter also converts [SvgGroup] to/from [GraphicComposite].
 *
 * @author elisha
 */
class SvgElementGraphicConverter : Converter<SvgElement?, Graphic<Graphics2D?>?>() {
    /**
     * Convert an [SvgElement] to a [Graphic]. The resulting object will
     * be a [GraphicComposite] if the argument is an instance of [SvgGroup]
     * or [SvgRoot], and otherwise a [PrimitiveGraphicSupport]. In the case of
     * a group, recursive calls are made to convert all elements in the group.
     *
     * @param sh the element to convert
     * @return the corresponding graphic
     */
    fun doForward(sh: SvgElement?): Graphic<Graphics2D?>? {
        var prim: Graphic<Graphics2D?>? = null
        val style = aggregateStyle(sh)
        if (sh is SvgRectangle) {
            val rsh: RectangularShape = SvgRectangle.Companion.shapeConverter().convert(sh as SvgRectangle?)
            prim = JGraphics.shape(rsh, style)
        } else if (sh is SvgEllipse) {
            val rsh: Shape = SvgEllipse.Companion.shapeConverter().convert(sh as SvgEllipse?)
            prim = JGraphics.shape(rsh, style)
        } else if (sh is SvgCircle) {
            val rsh: Shape = SvgCircle.Companion.shapeConverter().convert(sh as SvgCircle?)
            prim = JGraphics.shape(rsh, style)
        } else if (sh is SvgPolygon) {
            val rsh: Shape = SvgPolygon.Companion.shapeConverter().convert(sh as SvgPolygon?)
            prim = JGraphics.shape(rsh, style)
        } else if (sh is SvgLine) {
            val line: Line2D = SvgLine.Companion.shapeConverter().convert(sh as SvgLine?)
            prim = JGraphics.path(line, style)
        } else if (sh is SvgPolyline) {
            val rsh: Shape = SvgPolyline.Companion.shapeConverter().convert(sh as SvgPolyline?)
            prim = JGraphics.shape(rsh, style)
        } else if (sh is SvgPath) {
            val rsh: Shape = SvgPath.Companion.shapeConverter().convert(sh as SvgPath?)
            prim = JGraphics.shape(rsh, style)
        } else if (sh is SvgImage) {
            val img: AnchoredImage = SvgImage.Companion.imageConverter().convert(sh as SvgImage?)
            prim = JGraphics.image(img)
            prim.setMouseDisabled(true)
        } else if (sh is SvgText) {
            val text: AnchoredText = SvgText.Companion.textConverter().convert(sh as SvgText?)
            prim = JGraphics.text(text, style)
            prim.setMouseDisabled(true)
        } else if (sh is SvgGroup) {
            prim = GraphicComposite()
            (prim as GraphicComposite<*>?).setStyle(style)
            for (el in (sh as SvgGroup?).getElements()) {
                (prim as GraphicComposite<Graphics2D?>?).addGraphic(doForward(el))
            }
        } else {
            throw IllegalStateException("Unexpected Svg element: $sh")
        }
        prim.setDefaultTooltip(sh.id)
        return prim
    }

    private fun aggregateStyle(element: SvgElement?): AttributeSet? {
        val shapeStyle = element.getStyle()
        val res = shapeStyle?.copy() ?: AttributeSet()
        val attr = element.getOtherAttributes()
        if (attr != null) {
            for ((key, value) in attr) {
                val `val`: Any = AttributeSetCoder().decode(value as String)
                res.put(key.toString(), `val`)
            }
        }
        if (element.getId() != null) {
            res.put(Styles.ID, element.getId())
        }
        AttributeSetAdapter.Companion.updateColorFields(res)
        return res
    }

    fun doBackward(v: Graphic<Graphics2D?>?): SvgElement? {
        return graphicToSvg(v)
    }

    companion object {
        private val LOG = Logger.getLogger(SvgElementGraphicConverter::class.java.name)
        private val INST: SvgElementGraphicConverter? = SvgElementGraphicConverter()

        /**
         * Get global instance of the converter.
         * @return instance
         */
        fun getInstance(): Converter<SvgElement?, Graphic<Graphics2D?>?>? {
            return INST
        }

        /**
         * Convert a graphic component to an Svg object, including a view box.
         * @param compt component to convert
         * @return result
         */
        fun componentToSvg(compt: JGraphicComponent?): SvgRoot? {
            val root = SvgRoot()
            root.width = compt.getWidth().toDouble()
            root.height = compt.getHeight().toDouble()
            root.viewBoxAsRectangle = PanAndZoomHandler.Companion.getLocalBounds(compt)
            root.style.put("background", ColorCoder().encode(compt.getBackground()))
            root.style.put(Styles.FONT_SIZE, Styles.DEFAULT_TEXT_STYLE[Styles.FONT_SIZE])
            val group = getInstance().reverse()
                    .convert(compt.getGraphicRoot()) as SvgGroup
            group.elements.forEach(Consumer { obj: SvgElement? -> root.addElement(obj) })
            return root
        }
        //<editor-fold defaultstate="collapsed" desc="PRIVATE UTILITIES">
        /** Converts a graphic element to an Svg element  */
        private fun graphicToSvg(v: Graphic<Graphics2D?>?): SvgElement? {
            var res: SvgElement? = null
            res = if (v is LabeledShapeGraphic<*>) {
                labeledShapeToSvg(v as LabeledShapeGraphic<Graphics2D?>?)
            } else if (v is PrimitiveGraphicSupport<*, *>) {
                val pgs = v as PrimitiveGraphicSupport<*, *>?
                primitiveStyleToSvg(pgs.getPrimitive(), v.getStyle(), pgs.getRenderer())
            } else if (v is GraphicComposite<*>) {
                compositeToSvg(v as GraphicComposite<Graphics2D?>?)
            } else if (v is PrimitiveArrayGraphicSupport<*, *>) {
                primitiveArrayToSvg(v as PrimitiveArrayGraphicSupport<*, *>?)
            } else {
                throw IllegalArgumentException("Graphic conversion not supported for " + v.javaClass)
            }
            val id = v.renderStyle().getString(Styles.ID, null)
            if (id != null && res != null) {
                res.id = id
            }
            if (res != null && res.style != null && res.style.attributes.isEmpty()) {
                res.style = null
            }
            if (res != null && res.style != null) {
                for (c in res.style.attributes) {
                    val col = res.style[c]
                    if (col is Color) {
                        res.style.put(c, col as Color?. alpha 255)
                    }
                }
            }
            return res
        }

        /** Converts a blaise composite to an Svg group  */
        private fun compositeToSvg(gc: GraphicComposite<Graphics2D?>?): SvgElement? {
            val grp = SvgGroup()
            if (gc.getStyle() != null) {
                grp.style = AttributeSet.create(gc.getStyle().attributeMap)
            }
            for (g in gc.getGraphics()) {
                try {
                    val el = graphicToSvg(g)
                    if (el != null) {
                        grp.addElement(el)
                    } else {
                        LOG.log(Level.WARNING, "Null graphic for {0}", g)
                    }
                } catch (x: IllegalArgumentException) {
                    LOG.log(Level.WARNING, "Graphic not added to result", x)
                }
            }
            return grp
        }

        /** Converts a blaise array graphic to Svg group  */
        private fun primitiveArrayToSvg(pags: PrimitiveArrayGraphicSupport<*, *>?): SvgElement? {
            val grp = SvgGroup()
            grp.style = pags.renderStyle().flatCopy()
            for (o in pags.getPrimitive()) {
                grp.addElement(primitiveStyleToSvg(o, pags.renderStyle().flatCopy(), pags.getRenderer()))
            }
            return grp
        }

        /** Creates an Svg element from given primitive/style  */
        private fun primitiveStyleToSvg(primitive: Any?, sty: AttributeSet?, rend: Renderer<*, *>?): SvgElement? {
            return if (primitive is Shape) {
                SvgElements.create(null, primitive as Shape?, sty)
            } else if (primitive is AnchoredText) {
                SvgElements.create(null, primitive as AnchoredText?, sty, rend)
            } else if (primitive is AnchoredImage) {
                SvgElements.create(null, primitive as AnchoredImage?, sty)
            } else if (primitive is AnchoredIcon) {
                SvgElements.create(null, primitive as AnchoredIcon?, sty)
            } else if (primitive is Point2D) {
                SvgElements.create(null, primitive as Point2D?, sty)
            } else {
                throw IllegalArgumentException("Graphic conversion not supported for primitive $primitive")
            }
        }

        /** Converts a labeled shape to svg  */
        private fun labeledShapeToSvg(gfc: LabeledShapeGraphic<Graphics2D?>?): SvgElement? {
            val shape = primitiveStyleToSvg(gfc.getPrimitive(), gfc.renderStyle().flatCopy(), gfc.getRenderer())
            val text = labelToSvg(gfc)
            return if (text == null) shape else SvgGroup.Companion.create(shape, text)
        }

        /** Generates element for object label  */
        private fun labelToSvg(gfc: LabeledShapeGraphic<Graphics2D?>?): SvgElement? {
            val styler = gfc.getObjectStyler() ?: return null
            val label = styler.label(gfc.getSourceObject())
            val style = styler.labelStyle(gfc.getSourceObject())
            if (Strings.isNullOrEmpty(label) || style == null) {
                return null
            }
            val textRend = gfc.getTextRenderer()
            return if (textRend is WrappedTextRenderer) {
                SvgElements.createWrappedText(label, style, LabeledShapeGraphic.Companion.wrappedLabelBounds(gfc.getPrimitive()))
            } else if (textRend is TextRenderer) {
                primitiveStyleToSvg(AnchoredText(label), style.flatCopy(), textRend)
            } else {
                LOG.log(Level.WARNING, "Unsupported text renderer: {0}", textRend)
                null
            }
        } //</editor-fold>
    }
}